src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs

```diff
@@ -158,34 +158,48 @@ public virtual void SetParameters(ParameterCollection parameters)
                 _hasCalledInit = true;
                 OnInit();
 
-                // If you override OnInitAsync and return a nonnull task, then by default
+                // If you override OnInitAsync and return a noncompleted task, then by default
                 // we automatically re-render once that task completes.
                 var initTask = OnInitAsync();
-                if (initTask != null && initTask.Status != TaskStatus.RanToCompletion)
-                {
-                    initTask.ContinueWith(ContinueAfterLifecycleTask);
-                }
+                ContinueAfterLifecycleTask(initTask);
             }
 
             OnParametersSet();
             var parametersTask = OnParametersSetAsync();
-            if (parametersTask != null && parametersTask.Status != TaskStatus.RanToCompletion)
-            {
-                parametersTask.ContinueWith(ContinueAfterLifecycleTask);
-            }
+            ContinueAfterLifecycleTask(parametersTask);
 
             StateHasChanged();
         }
 
-        private void ContinueAfterLifecycleTask(Task task)
+        private async void ContinueAfterLifecycleTask(Task task)
         {
-            if (task.Exception == null)
+            if (task == null)
             {
-                StateHasChanged();
+                // Treat like a pre-completed task that can't possibly raise errors
+                // i.e., do nothing
+            }
+            else if (task.IsCompleted)
+            {
+                // Since it's already completed synchronously, no need to await and no
+                // need to issue a further render (we already rerender synchronously).
+                // Just need to make sure we propagate any errors.
+                if (task.IsFaulted)
+                {
+                    HandleException(task.Exception);
+                }
             }
             else
             {
-                HandleException(task.Exception);
+                // For incomplete tasks, automatically re-render on completion
+                try
+                {
+                    await task;
+                    StateHasChanged();
+                }
+                catch (Exception ex)
```

Comment:
There's an inconsistency here. A synchronously cancelled cancelled task will have its exception ignored - since `Task.IsCompleted` includes all three terminal states. https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscompleted?view=netframework-4.7.2

But an asynchronously cancelled task will have its exception reported.

IMO it would be good to skip reporting exceptions when a task is cancelled. Cancellation is usually the result of a shutdown or dispose in UI apps and you don't really need to know about it.

Reply:
It might be slightly more clear to implement the synchronous logic as a switch over https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskstatus?view=netframework-4.7.2

Reply:
Very good point. Done.